### PR TITLE
Allow dynamical decoupling pass to work on circuits with pulse gates 

### LINKED ISSUE #10833: Dynamical decoupling fails when a gate in the decoupling sequence has a calibration attached to the circuit
### Environment

- **Qiskit Terra version**: 0.25.1
- **Python version**: 3.10
- **Operating system**: Fedora Linux

### What is happening?

`PadDynamicalDecoupling` raises `TypeError: unhashable type: 'list'` when run on a circuit with a calibration for one of the gates in the decoupling sequence.

### How can we reproduce the issue?

```python
from qiskit_ibm_provider import IBMProvider


provider = IBMProvider()
backend = provider.get_backend("ibm_nazca")


import numpy as np

from qiskit import pulse, QuantumCircuit
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import ALAPScheduleAnalysis, PadDynamicalDecoupling
from qiskit.circuit.library import XGate, YGate
from qiskit.pulse import DriveChannel


inst_sched_map = backend.instruction_schedule_map


circ = QuantumCircuit(1, 1)
circ.x(0)
circ.delay(1600, 0)
circ.x(0)

for qubit in range(backend.configuration().n_qubits):
    with pulse.build(f"y gate for qubit {qubit}") as sched:
        with pulse.phase_offset(np.pi / 2, DriveChannel(qubit)):
            x_sched = inst_sched_map.get("x", qubits=[qubit])
            pulse.call(x_sched)
        circ.add_calibration("y", [qubit], sched)

dd_sequence = [XGate(), YGate()]*4
durations = backend.target.durations()

pm = PassManager(
        [
            ALAPScheduleAnalysis(durations),
            PadDynamicalDecoupling(
                durations,
                dd_sequence,
                pulse_alignment=backend.configuration().timing_constraints[
                    "pulse_alignment"
                ],
            ),
        ]
    )

pm.run(circ)
```

### What should happen?

The duration of the gate calibration should be used to schedule the decoupling sequence.

### Any suggestions?

A PR with a fix will be opened shortly.

### PR DESCRIPTION
**Pull request by wshanks**
### Summary

Allow dynamical decoupling pass to work on circuits with pulse gates 

### Details and comments

The pass was accessing the circuit calibrations incorrectly, resulting in
an error when one of the gates in the dynamical decoupling sequence had
a pulse gate calibration.

Closes #10833

### CODE REVIEW

**Comment by wshanks** [comment]
Closes #10833

File: qiskit/transpiler/passes/scheduling/padding/dynamical_decoupling.py
Comment (jakelishman) [inline qiskit/transpiler/passes/scheduling/padding/dynamical_decoupling.py]: In `QuantumCircuit.has_calibration_for`, the `params` tuple is built by coercing bound `ParameterExpression` objects to `float` as well: https://github.com/Qiskit/qiskit/blob/e5997133d6d8c89405c17e89117216e986a87462/qiskit/circuit/quantumcircuit.py#L451-L456

Does that need to be done here too?

**Review by jakelishman — COMMENTED** [review]
In principle this looks fine to me, though I'd wait for one of the pulse maintainers to review properly.

**Comment by jakelishman** [comment]
You need to put the "closes #10833" in the top PR comment for it to have an effect, but you can edit it in and it'll work.

**Comment by wshanks** [comment]
Ah, okay. I put it separately because I don't like it ending up in the commit message, but I guess it's fine.

File: qiskit/transpiler/passes/scheduling/padding/dynamical_decoupling.py
Comment (wshanks) [reply qiskit/transpiler/passes/scheduling/padding/dynamical_decoupling.py]: I think you are right. I think it would probably be best to have a `QuantumCircuit.get_calibration(gate, qubits, params)` method to abstract this operation, but I won't add that here.

File: test/python/transpiler/test_dynamical_decoupling.py
Comment (nkanazawa1989) [inline test/python/transpiler/test_dynamical_decoupling.py:20]: Imported twice

File: test/python/transpiler/test_dynamical_decoupling.py
Comment (nkanazawa1989) [inline test/python/transpiler/test_dynamical_decoupling.py:342]: Do you also want a test for parametric echo gate? I don't think this is realistic and it's up to you.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks Will. This looks good to me. I'll approve and add this to queue once linter error is fixed.

File: qiskit/transpiler/passes/scheduling/padding/dynamical_decoupling.py
Comment (jakelishman) [reply qiskit/transpiler/passes/scheduling/padding/dynamical_decoupling.py]: Yeah, agreed on both parts.

**Comment by jakelishman** [comment]
> I put it separately because I don't like it ending up in the commit message, but I guess it's fine.

The PR body isn't used for the commit message in a squash-merge - GitHub just catenates all the commit messages together, so if you've got stuff that's added to the PR comment but not the head commit message, it doesn't get included.  (I think the rules are different if the PR is merged by merge, though.)

File: test/python/transpiler/test_dynamical_decoupling.py
Comment (wshanks) [reply test/python/transpiler/test_dynamical_decoupling.py:20]: Fixed!

File: test/python/transpiler/test_dynamical_decoupling.py
Comment (wshanks) [reply test/python/transpiler/test_dynamical_decoupling.py:342]: Since the code handles parameters, we might as well test them. I added an extra test.

**Comment by wshanks** [comment]
> The PR body isn't used for the commit message in a squash-merge - GitHub just catenates all the commit messages together, so if you've got stuff that's added to the PR comment but not the head commit message, it doesn't get included. (I think the rules are different if the PR is merged by merge, though.)

Ah, okay, I have not looked closely at the commit messages since this repo switched to using the merge queue (EDIT: actually, I see now that mergify was also configured for this commit message format, but back then one could also do a direct "Squash and merge" though one was not supposed to because it require extra CI churn). In qiskit-experiments, when we started using the merge queue, we changed the default commit message to "Pull request title and body" so that we could have more control over the commit message. I forgot that wasn't the default behavior. I hope some day the "Merge when ready" button will open a text box for editing the commit message like "Squash and merge" does.

Now I want to squash and force push before merging :thinking:

**Review by nkanazawa1989 — APPROVED** [review]
Thanks. This looks good to me.

**Pull request merged**
